src/verifier/mod.rs

```diff
@@ -1,120 +1,83 @@
 #[derive(Debug, PartialEq, Eq)]
-pub enum GuessChar {
-    Hit(char),
-    Miss(char),
-    Move(char),
+pub enum GuessResult {
+    Guessed,
+    Missed { hint: String },
 }
 
 #[derive(Debug)]
-pub struct Verifier {
-    input: Vec<GuessChar>,
-    secret: Vec<char>,
+pub struct Verifier<'a> {
+    secret: &'a str,
```

Comment:
Having a reference here in the struct, can be a good exercise of lifetimes :)
In practice, however, you should always take a pgramatical trade off:
* Either you trying to squeeze maximum performance and avoid any extra memory allocation and take the hard maintenance burden (yeah, if you have quickly evolving code base, lifetimes may become a pain in a neck)
* Or you just make an extra allocation and go on with it.

Reply:
Agree with you, @greyblake  I've overcomplicated it.
I made it with learning purposes, but yes, the owned type `String` suits here the best and does not imply any performance impact.

Reply:
applied changes: 167ac962b9ff8aae9209feacb6429b6b2b019852